### ISSUE TITLE
Move away from `@ng-dynamic-forms`

### ISSUE THREAD

**Issue by ybnd** [DSpace/dspace-angular#2216]
#### Issue / context

During the Angular 15 upgrade, @Atmire-Kristof noticed conflicts between the newer versions of the following libraries:

* [`[ng-bootstrap](https://github.com/ng-bootstrap/ng-bootstrap)`](https://github.com/ng-bootstrap/ng-bootstrap)
  * v14 supports Angular 15 properly
* [`[@ng-dynamic-forms](https://github.com/udos86/ng-dynamic-forms)`](https://github.com/udos86/ng-dynamic-forms)
  * v16+ claims to support Angular 15, but `@ng-dynamic-forms/ui-ng-bootstrap` still requires `ng-bootstrap` v11 as a peer dependency, which itself _doesn't support Angular 15_
  * After "forcing" the upgrade, builds fail because of inconsistencies between the `ng-bootstrap` versions, so this is not a viable workaround

Rather than do a half-upgrade we opted to leave both of these libraries as-is for now. This leads to peer dependency mismatch warnings, but everything remains functional.

#### Solution

It looks like development of `@ng-dynamic-forms` has slowed down as of late, and this version mismatch is a bit of a red flag. Therefore, **we should consider moving away from this library and use the default Angular forms instead** for more stability.

This would involve a significant refactor of DSpace Angular submission.

#### References

_Add references/links to any related issues or PRs. These may include:_

* Related to #858

* Left out of #2204

**Comment by saschaszott** [comment]
related issue: #3226

**Comment by atarix83** [comment]
Regarding this topic I'd propose to integrate specific parts of the ng-dynamic-forms library into the DSpace base code. Specifically, we should port the [core](https://github.com/udos86/ng-dynamic-forms/tree/master/projects/ng-dynamic-forms/core) and [ui-ng-bootstrap](https://github.com/udos86/ng-dynamic-forms/tree/master/projects/ng-dynamic-forms/ui-ng-bootstrap) modules.

While @artlowel recently suggested migrating to [ngx-formly](https://github.com/ngx-formly/ngx-forml), it’s important to note that ngx-formly is still based on Angular 13. In contrast, the current version of ng-dynamic-forms is built on Angular 16 and utilizes standalone components, which is advantageous for bundle size. Although ng-dynamic-forms requires updates to the latest versions of Angular and Bootstrap/ng-bootstrap, we (4science) have already initiated this process with a [pull request](https://github.com/udos86/ng-dynamic-forms/pull/1185).

By incorporating the necessary parts of ng-dynamic-forms into the base code, we can minimize the workload associated with migrating to a different library. This approach will also simplify code maintenance and version management.

This work could be done in two different phases:

**Phase 1**: Initial Integration, in order to move the necessary parts of ng-dynamic-forms into the DSpace base code.
  - Identify and extract the required modules from ng-dynamic-forms (core and ui-ng-bootstrap).
  - Update component references in the existing codebase to point to the newly integrated modules.
  -  Remove the ng-dynamic-forms dependency from the project.

Estimated Time: 2 days

**Phase 2**: Refactor the DSpace code that currently uses a custom extension of ui-ng-bootstrap https://github.com/DSpace/dspace-angular/tree/main/src/app/shared/form/builder/ds-dynamic-form-ui.
  - Review the custom extension of ui-ng-bootstrap in the DSpace code (ds-dynamic-form-ui).
  - Refactor the code to align with the integrated ng-dynamic-forms modules.
  - Ensure all custom functionalities are preserved and optimized.
  - Conduct thorough testing to verify the refactored code works as expected.

Estimated Time: 4 days.

**Comment by artlowel** [comment]
>While @artlowel recently suggested migrating to [ngx-formly](https://github.com/ngx-formly/ngx-forml), it’s important to note that ngx-formly is still based on Angular 13.

The minimal required version of angular is version 13, but it is compatible with angular 17, as you can tell from [my proof of concept](https://github.com/atmire/dspace-angular/tree/formly-test). It builds without peer dependency warnings. It is likely also compatible with 18, but I haven't verified that myself yet. Its dependencies were last updated last week, so I'm sure if there were breaking changes they'd make a dedicated version for angular 18

>By incorporating the necessary parts of ng-dynamic-forms into the base code, we can minimize the workload associated with migrating to a different library. This approach will also simplify code maintenance and version management.

There are very few DSpace developers that understand how to even _use_ the ng-dynamic-forms library well enough to make meaningful changes to the submission. If we were to include it into our codebase they'd have to learn to maintain the lib itself too. That's just not something that is likely to happen.

It would also be a big PR that we couldn't really review properly, but would just have to merge and hope for the best. 

We should take this opportunity to move away from that lib, and rewrite the submission into something that is easier to understand, modify and maintain.

It doesn't have to be formly, as discussed, I'm also working on a PoC using Angular 18 forms, but I'm convinced sticking with ng-dynamic-forms is a bad idea

**Comment by artlowel** [comment]
>Even if we start from scratch, the new code would still need to be maintained, and other developers would have to learn it to ensure proper maintenance

But we could design it to be much easier to understand and much easier to maintain. It currently isn't, and bringing in that entire lib is only going to exacerbate that problem.

Of course it would be less work right now. But we'd be pushing the problem of the refactor we've known has been necessary for years down the road yet again, taking on a bunch of extra code to maintain in the process. By going for the quick fix now, the submission will remain hard to modify, debug and maintain, which will end up costing us a lot more effort in the long run.

>By including the library in our codebase, we would have a working solution that could serve as a solid foundation for improvements, especially considering the new features available in Angular 18, such as the unified control state change events combined with Signals

This lib is only going to get support for angular 18 and things like "unified control state change events combined with signals" if we add it ourselves. So we might as well do that without that lib in the first place, or get it in the next version of formly

**Comment by atarix83** [comment]
Since the submission form is closely related to the migration of the @ng-dynamic-forms library, I am compiling a list of functionalities that should be supported by the new library to ensure the submission form continues to function as it currently does:
- **Submission Metadata Enhancement**: The submission form should be able to automatically enhance metadata by importing it from external providers (e.g., Scopus, PubMed). This can be triggered by an identifier or a PDF upload. These functionalities are present in DSpace-CRIS for REST implementation and are fully supported by the current DSpace Angular application.
[Video del 2024-11-29 16-23-55.webm](https://github.com/user-attachments/assets/9ff7a470-d08f-4e4d-8bc7-69a5355a469a)
- **Server-Side Validation**: It should be possible to implement custom logic validators on the server side, which trigger errors to be displayed in the submission form. These validators do not need to be known when the form is initialized but can use the errors of the workspace/workflow item objects, for example:
  ```
    "errors" : [ {
      "message" : "error.validation.server-side-custom",
      "paths" : [ "/sections/tratitional/dc.title" ]
    }, {
      "message" : "error.validation.license.required",
      "paths" : [ "/sections/license" ]
    } ],
  ```
- **Metadata Authority Confidence**: The submission form should use a system to handle the authority confidence level, displaying a status icon according to the value of the authority confidence.
  ![Schermata del 2024-11-29 16-34-08](https://github.com/user-attachments/assets/ce96f378-cf94-4103-85c9-dae62d1d1fcf)
- **Repeatable Metadata**: The submission form should allow to handle the order of metadata values using drag and drop 
  ![image](https://github.com/user-attachments/assets/0ab17c19-faf6-4041-9294-aa892f863468)
- **Type bind**: The submission form should allow to control the visibility of form fields based on the type of item being submitted.

**Comment by saschaszott** [comment]
As part of the replacement of ng-dynamic-forms, we should also take a look at Angular Signal Forms (currently still experimental).

**Comment by artlowel** [comment]
We discussed this in the meeting today. Here's a summary of the status so far:

- I made a proof of concept using [ngx-formly](https://github.com/ngx-formly/ngx-formly), that replaces the default submission form with a new one that only renders oneboxes and dates for now, just to get a feel for the library.
- When we discussed that proof of concept in a meeting it was suggested that we should compare it to a solution using the native forms in angular 18 which had gotten some improvements
- So I made a second prototype doing that also implements a onebox and a date field using angular 18 forms, without an extra form library.
- The main reason I put this work on hold was because of the CRIS discussion. No matter the outcome, there would not be enough time to refactor the entire submission form in dspace 10.
- My conclusion at the time was that the pure angular solution was probably the way to go. While it was a little more work, it does give us the option to completely tailor it to DSpace. 
	- For example the form fields in the angular 18 prototype [work directly with MetadataValues](https://github.com/atmire/ng18-submission-poc/blob/main/libs/ng18-submission/src/submission/form-fields/date-field/date-field.component.ts#L81-L102), there's no need for an in-between format that then needs to be translated every time we send or receive messages from the server
- The biggest downside to using pure angular forms, is that there's a lot more freedom, so also the freedom to make poor choices. We'll need to come up with good conventions around how to design a form field, document them well, and enforce them.
- You can find the prototypes here: 
	- formly: https://github.com/atmire/dspace-angular/tree/formly-test
	- Angular 18: https://github.com/atmire/ng18-submission-poc/
- Here's the same date field implemented in both ways, for comparison:
	- formly: https://github.com/atmire/dspace-angular/tree/formly-test/src/app/submission-2/submission-page/date-field
	- Angular 18: https://github.com/atmire/ng18-submission-poc/tree/main/libs/ng18-submission/src/submission/form-fields/date-field

**Comment by strawburster** [comment]
I agree with that reasoning, I think angular forms may be the way to go here. I'm curious if you could expand, @artlowel  on your opinion that the greater freedom would result in more ability to make poor choices. In my view, looking over your examples, the complicated bits are all in the custom field components. If we pay special attention to those components, we should be able to reuse them throughout the project without too many degrees of freedom. Do you have concerns for work on the field components themselves, or the forms in general?

**Comment by artlowel** [comment]
@strawburster You're right, the complicated bits are those custom field components. Angular doesn't force you to implement those in any particular way. For example, while it's good practice to implement the ControlValueAccessor and Validator interfaces, you're not required to do so. There are also no rules for the format of the value a custom form field returns. So we'll need good guidelines on how to write a custom form field in DSpace, and ideally also have some way to detect form fields in PRs that don't adhere to those guidelines, for example using lint rules.
